opensrp-fp/src/main/java/org/smartregister/fp/common/task/FinalizeVisitFormTask.java

```diff
@@ -0,0 +1,174 @@
+package org.smartregister.fp.common.task;
+
+import android.app.ProgressDialog;
+import android.content.Context;
+import android.os.AsyncTask;
+import android.text.TextUtils;
+
+import com.vijay.jsonwizard.constants.JsonFormConstants;
+import com.vijay.jsonwizard.rules.RuleConstant;
+
+import org.joda.time.LocalDate;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.smartregister.fp.R;
+import org.smartregister.fp.common.domain.Contact;
+import org.smartregister.fp.common.domain.WomanDetail;
+import org.smartregister.fp.common.library.FPLibrary;
+import org.smartregister.fp.common.repository.PreviousContactRepository;
+import org.smartregister.fp.common.rule.ContactRule;
+import org.smartregister.fp.common.util.ConstantsUtils;
+import org.smartregister.fp.common.util.DBConstantsUtils;
+import org.smartregister.fp.common.util.FPFormUtils;
+import org.smartregister.fp.common.util.Utils;
+import org.smartregister.fp.features.home.repository.PatientRepository;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import timber.log.Timber;
+
+import static org.smartregister.fp.common.util.Utils.isCheckboxValueEmpty;
+
+public class FinalizeVisitFormTask extends AsyncTask<Void, Void, Void> {
+
+    private final Context context;
+    private final String baseEntityId;
+    private final int contactNo;
+    private final Contact contact;
+    private final String jsonCurrentState;
+    private final FPFormUtils fpFormUtils;
+    private final ProgressDialog progressDialog;
+    private HashMap<String, String> clientProfileDetail;
+
+    public FinalizeVisitFormTask(Context context, String baseEntityId, int contactNo, Contact contact, String jsonCurrentState) {
+        this.context = context;
+        this.baseEntityId = baseEntityId;
+        this.contactNo = contactNo;
+        this.contact = contact;
+        this.jsonCurrentState = jsonCurrentState;
+        fpFormUtils = new FPFormUtils();
+
+        progressDialog = new ProgressDialog(context);
```

Comment:
Display the progress dialog on the method `onPreExecute()`.  You can construct the object in the constructor but run `.show()` on PreExecute. The benefit to this is that running you are not left with a progress dialog rotating and not dismissed in the event the AsyncTask runs into an error during construction.

Reply:
yeah, that's the good point that I have just forgotten. done that